services/command-router-base/src/main/java/org/eclipse/hono/commandrouter/impl/KubernetesBasedAdapterInstanceStatusService.java

```diff
@@ -0,0 +1,344 @@
+/*******************************************************************************
+ * Copyright (c) 2021 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+
+package org.eclipse.hono.commandrouter.impl;
+
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.regex.Matcher;
+
+import org.eclipse.hono.commandrouter.AdapterInstanceStatusService;
+import org.eclipse.hono.util.AdapterInstanceStatus;
+import org.eclipse.hono.util.CommandConstants;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.fabric8.kubernetes.api.model.ContainerStatus;
+import io.fabric8.kubernetes.api.model.Pod;
+import io.fabric8.kubernetes.client.DefaultKubernetesClient;
+import io.fabric8.kubernetes.client.KubernetesClient;
+import io.fabric8.kubernetes.client.KubernetesClientException;
+import io.fabric8.kubernetes.client.Watch;
+import io.fabric8.kubernetes.client.Watcher;
+import io.fabric8.kubernetes.client.WatcherException;
+import io.vertx.core.Future;
+
+/**
+ * A service for determining the status of adapter instances that run in a Kubernetes cluster along with the component
+ * this service is running in.
+ * <p>
+ * Note that this component must be allowed to "list" and "watch" pods, requiring a corresponding RBAC configuration.
+ */
+public class KubernetesBasedAdapterInstanceStatusService implements AdapterInstanceStatusService {
+
+    private static final Logger LOG = LoggerFactory.getLogger(KubernetesBasedAdapterInstanceStatusService.class);
+    private static final String ADAPTER_NAME_MATCH = "adapter";
+    private static final long WATCH_RECREATION_DELAY_MILLIS = 100;
+    private static final int MAX_LRU_MAP_ENTRIES = 200;
+
+    private final KubernetesClient client;
+    private final String namespace;
+    /**
+     * Values: 0: inactive, 1: active, -1: status service stopped.
+     */
+    private final AtomicInteger active = new AtomicInteger();
```

Comment:
why do we need do differentiate three states? If the service needs to be started before it can be used, then FMPOV we would simply throw an IllegalStateException if somebody tries to invoke its methods before it has been started.

Reply:
A switch between inactive and active is also done when the Watcher has been closed and is then re-initiated.
Apart from that there is the state of the whole service having been closed.
So, the `getStatus` method has to check both whether the service is currently inactive (Watcher being re-initialized) or stopped. In order not to check two Booleans here, I've chosen the AtomicInteger, to be updated atomically.
See also the other places where the `active` flag is used (e.g. Watcher re-init should be aborted if service already closed).